### ISSUE TITLE
Add VxLAN Support

### ISSUE THREAD

**Issue by yicwang** [uni-tue-kn/P4TG#8]
Hi,

This is a feature request. :)

I see we have a lot of encapsulations supported, but not VxLAN. Actually VxLAN is so a common encapsulation in data centers. Basically we will add the OUTER SMAC, OUTER DMAC, OUTER IP SRC, OUTER IP DST, OUTER SPORT, OUTER DPORT (4789), OUTER PROTO (UDP), VNID, and VXLAN FLAGS (as plus) for egress, and remove this headers for ingress.

Most of the fields can be made to be configurable, just like INNER frames. In my case, my traffic is 100% VxLAN, so I did a very hacky implementation just to meet my deadline:
https://github.com/yicwang/P4TG/commit/b6553b2eb3418314a49ec81c8bcfd59f7b0831e3
Only challenge is we have to deal with the checksum correctly. 

I am very happy to contribute a formal VxLAN encap/decap implementation when I've got some time.

Thanks very much!

Regards,
Yichen

**Comment by NE4Y** [comment]
Hi,

good idea. I just looked at the VxLAN RFC and it seems reasonable to support it.
Thanks for the preliminary implementation!

I'd like to implement it as with the other encapsulation methods, i.e., the generated packet is already a VxLAN packet and the dataplane rewrites its headers.

Could you explain the "remove this header for ingress" part? Is this for the "analyze" mode, i.e., when P4TG only forwards traffic, or also for the generation mode?

Best

Steffen

**Comment by yicwang** [comment]
Hi Steffen,

Yes, generated and rewrite is a better idea!

It is for analyzing. As in today's ingress code, we do parse out all the encapsulations (VLAN/QinQ/MPLS) because we want to see the actual UDP payload with the P4TG header. Same thing that is needed for VxLAN as well. For VxLAN it is tricky because you will need to parse more layers to reach the final payload. Something like:

```p4
    state parse_ethernet {
        pkt.extract(hdr.ethernet);
        transition select(hdr.ethernet.ether_type) {
            ETHERTYPE_IPV4 : parse_ipv4;
            ETHERTYPE_ARP: parse_arp;
            default : reject;
        }
    }

    state parse_ipv4 {
        pkt.extract(hdr.ipv4);
        transition select(hdr.ipv4.protocol) {
            IP_PROTOCOL_TCP : parse_tcp;
            IP_PROTOCOL_UDP : parse_udp;
            default : reject;
        }
    }

    state parse_udp {
        pkt.extract(hdr.udp);
        transition select(hdr.udp.dst_port) {
            UDP_PORT_VXLAN: parse_vxlan;
            default: accept;
        }
    }

    state parse_vxlan {
        pkt.extract(hdr.vxlan);
        transition parse_inner_ethernet;
    }

    state parse_inner_ethernet {
        pkt.extract(hdr.inner_ethernet);
        transition select(hdr.inner_ethernet.ether_type) {
            ETHERTYPE_IPV4 : parse_inner_ipv4;
            ETHERTYPE_IPV6 : parse_inner_ipv6;
            default: reject;
        }
    }

    state parse_inner_ipv4 {
        pkt.extract(hdr.inner_ipv4);
        transition select(hdr.inner_ipv4.protocol) {
            IP_PROTOCOL_TCP : parse_inner_tcp;
            IP_PROTOCOL_UDP : parse_inner_udp;
            default : reject;
        }
    }

    state parse_inner_udp {
        pkt.extract(hdr.inner_udp);
        transition accept;
    }
``` 

Regards,
Yichen

**Comment by NE4Y** [comment]
Ah okay, gotcha.

Its on the list, should not take too long 👍🏼

**Comment by NE4Y** [comment]
A first version with VxLAN support is available at acf07de on the VxLAN branch.
<img width="1505" alt="Bildschirmfoto 2024-03-07 um 14 08 45" src="https://github.com/uni-tue-kn/P4TG/assets/7404984/69d3c443-6a0f-4ac6-b154-9b46a4e797b4">
<img width="1708" alt="Bildschirmfoto 2024-03-07 um 14 08 59" src="https://github.com/uni-tue-kn/P4TG/assets/7404984/c4f12ebb-6fd9-48fc-8130-1a122a76d9fb">

It is possible to combine VxLAN tunnel with all available encapsulations (VLAN, QinQ, MPLS).
The outer UDP checksum is set to zero as recommended in RFC 7348, but the inner UDP checksum (from the P4TG header) is still updated correctly.

Would be great if you could test it if it works on your side too.

**Comment by yicwang** [comment]
You got this done so fast... The original controller crashed with my VxLAN P4 code:
[2024-03-07T23:27:25Z INFO  p4tg_controller::core::traffic_gen] Calculated traffic generation for stream #1. #1 packets per 8352 ns. Rate: 1 Gbps. Accuracy: 100.00%.
thread 'tokio-runtime-worker' panicked at src/core/traffic_gen.rs:511:59:
called `Option::unwrap()` on a `None` value
, and not sure what was the problem. Was debugging that and saw your new feature already released, so I gave it a try.

So I loaded the controller, initially the "settings" page is not loading in Chrome. So I changed to Firefox, and it loaded fine. The feature works as expected! Thanks a lot!

During testing, I found another good thing to have, will open other ticket.

Regards,
Yichen

**Comment by NE4Y** [comment]
Great to hear!

The blank settings page is probably due to the change of the stream description (now includes vxlan) and your old stored description in the local storage of your browser.

After you delete your local storage and reload the page, it should be re-initialized correctly.

Will be merged into the dev branch and after some more testing and enhancements into main.

**Comment by NE4Y** [comment]
Merged in ed6605db8decac194bd35c83507fb695f50043a6 to dev.
